File: src/reconciliation/TwoPointerReconciler.cpp
#include "reconciliation/TwoPointerReconciler.hpp"
#include "reconciliation/MatchClassifier.hpp"

#include <algorithm>

namespace reconciliation::engine {

using models::Transaction;

namespace {

struct KeyedRef {
    std::string key;
    const Transaction* tx;
};

// Returns the index one past the last element sharing keyed[start].key,
// i.e. [start, end) is the run of equal-key elements. Requires `keyed` to
// already be sorted by key.
std::size_t groupEnd(const std::vector<KeyedRef>& keyed, std::size_t start) {
    std::size_t end = start + 1;
    while (end < keyed.size() && keyed[end].key == keyed[start].key) {
        end++;
    }
    return end;
}

} // namespace

std::vector<ReconciliationResult> TwoPointerReconciler::reconcile(
        const std::vector<Transaction>& internalTransactions,
        const std::vector<Transaction>& externalTransactions,
        const MatchingConfig& config) const {

    std::vector<KeyedRef> internalKeyed;
    internalKeyed.reserve(internalTransactions.size());
    for (const auto& t : internalTransactions) {
        internalKeyed.push_back({MatchClassifier::buildKey(t, config.strategy), &t});
    }
    std::vector<KeyedRef> externalKeyed;
    externalKeyed.reserve(externalTransactions.size());
    for (const auto& t : externalTransactions) {
        externalKeyed.push_back({MatchClassifier::buildKey(t, config.strategy), &t});
    }

    // The O(N log N + M log M) step. Sorting by key is what makes the
    // subsequent merge-walk a single O(N + M) linear pass instead of
    // needing to search.
    auto byKey = [](const KeyedRef& a, const KeyedRef& b) { return a.key < b.key; };
    std::sort(internalKeyed.begin(), internalKeyed.end(), byKey);
    std::sort(externalKeyed.begin(), externalKeyed.end(), byKey);

    std::vector<ReconciliationResult> results;
    results.reserve(internalKeyed.size() + externalKeyed.size());

    std::size_t i = 0, j = 0;
    const std::size_t n = internalKeyed.size(), m = externalKeyed.size();

    while (i < n || j < m) {
        bool haveInternal = i < n;
        bool haveExternal = j < m;

        if (haveInternal && (!haveExternal || internalKeyed[i].key < externalKeyed[j].key)) {
            // Internal-only key at this point in the merge.
            std::size_t end = groupEnd(internalKeyed, i);
            std::size_t groupSize = end - i;
            for (std::size_t k = i; k < end; ++k) {
                results.push_back(groupSize > 1
                    ? MatchClassifier::duplicate(*internalKeyed[k].tx)
                    : MatchClassifier::missingExternal(*internalKeyed[k].tx));
            }
            i = end;

        } else if (haveExternal && (!haveInternal || externalKeyed[j].key < internalKeyed[i].key)) {
            // External-only key.
            std::size_t end = groupEnd(externalKeyed, j);
            std::size_t groupSize = end - j;
            for (std::size_t k = j; k < end; ++k) {
                results.push_back(groupSize > 1
                    ? MatchClassifier::duplicate(*externalKeyed[k].tx)
                    : MatchClassifier::missingInternal(*externalKeyed[k].tx));
            }
            j = end;

        } else {
            // Equal keys on both sides.
            std::size_t iEnd = groupEnd(internalKeyed, i);
            std::size_t jEnd = groupEnd(externalKeyed, j);
            std::size_t internalGroupSize = iEnd - i;
            std::size_t externalGroupSize = jEnd - j;

            if (internalGroupSize > 1 || externalGroupSize > 1) {
                // See HashReconciler.cpp for the full rationale: both
                // sides' records under this key are reported as duplicate
                // once EITHER side is ambiguous, so a size-1 side under an
                // ambiguous key is never silently dropped.
                for (std::size_t k = i; k < iEnd; ++k) {
                    results.push_back(MatchClassifier::duplicate(*internalKeyed[k].tx));
                }
                for (std::size_t k = j; k < jEnd; ++k) {
                    results.push_back(MatchClassifier::duplicate(*externalKeyed[k].tx));
                }
            } else {
                results.push_back(MatchClassifier::classifyMatch(
                    *internalKeyed[i].tx, *externalKeyed[j].tx, config));
            }

            i = iEnd;
            j = jEnd;
        }
    }

    return results;
}

} // namespace reconciliation::engine

File: recon_engine/reconciliation/two_pointer_reconciler.py
"""
TwoPointerReconciler
----------------------
Sorts both sides by key (the O(N log N + M log M) step), then does a single
linear merge-walk (O(N + M)) comparing keys pairwise, exactly like the merge
step of merge sort. Sorting is what makes the merge-walk a single linear
pass instead of needing to search.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from ..models.transaction import Transaction
from . import match_classifier as mc
from .base import IReconciler
from .config_and_result import MatchingConfig, ReconciliationResult


@dataclass
class _KeyedRef:
    key: str
    tx: Transaction


def _group_end(keyed: List[_KeyedRef], start: int) -> int:
    """Returns the index one past the last element sharing keyed[start].key,
    i.e. [start, end) is the run of equal-key elements. Requires `keyed` to
    already be sorted by key."""
    end = start + 1
    while end < len(keyed) and keyed[end].key == keyed[start].key:
        end += 1
    return end


class TwoPointerReconciler(IReconciler):
    @property
    def name(self) -> str:
        return "two_pointer"

    def reconcile(
        self,
        internal_transactions: List[Transaction],
        external_transactions: List[Transaction],
        config: MatchingConfig,
    ) -> List[ReconciliationResult]:
        internal_keyed = [_KeyedRef(mc.build_key(t, config.strategy), t) for t in internal_transactions]
        external_keyed = [_KeyedRef(mc.build_key(t, config.strategy), t) for t in external_transactions]

        # The O(N log N + M log M) step.
        internal_keyed.sort(key=lambda kr: kr.key)
        external_keyed.sort(key=lambda kr: kr.key)

        results: List[ReconciliationResult] = []
        i = j = 0
        n, m = len(internal_keyed), len(external_keyed)

        while i < n or j < m:
            have_internal = i < n
            have_external = j < m

            if have_internal and (not have_external or internal_keyed[i].key < external_keyed[j].key):
                # Internal-only key at this point in the merge.
                end = _group_end(internal_keyed, i)
                group_size = end - i
                for k in range(i, end):
                    results.append(
                        mc.duplicate(internal_keyed[k].tx) if group_size > 1
                        else mc.missing_external(internal_keyed[k].tx)
                    )
                i = end

            elif have_external and (not have_internal or external_keyed[j].key < internal_keyed[i].key):
                # External-only key.
                end = _group_end(external_keyed, j)
                group_size = end - j
                for k in range(j, end):
                    results.append(
                        mc.duplicate(external_keyed[k].tx) if group_size > 1
                        else mc.missing_internal(external_keyed[k].tx)
                    )
                j = end

            else:
                # Equal keys on both sides.
                i_end = _group_end(internal_keyed, i)
                j_end = _group_end(external_keyed, j)
                internal_group_size = i_end - i
                external_group_size = j_end - j

                if internal_group_size > 1 or external_group_size > 1:
                    # See hash_reconciler.py for the full rationale: both
                    # sides' records under this key are reported as
                    # duplicate once EITHER side is ambiguous, so a size-1
                    # side under an ambiguous key is never silently dropped.
                    for k in range(i, i_end):
                        results.append(mc.duplicate(internal_keyed[k].tx))
                    for k in range(j, j_end):
                        results.append(mc.duplicate(external_keyed[k].tx))
                else:
                    results.append(mc.classify_match(internal_keyed[i].tx, external_keyed[j].tx, config))

                i, j = i_end, j_end

        return results
